src/Storages/StorageMergeTree.cpp

```diff
@@ -2181,6 +2181,35 @@ std::unique_ptr<MergeTreeSettings> StorageMergeTree::getDefaultSettings() const
     return std::make_unique<MergeTreeSettings>(getContext()->getMergeTreeSettings());
 }
 
+PreparedSetsCachePtr StorageMergeTree::getPreparedSetsCache(Int64 mutation_id)
+{
+    auto l = std::lock_guard(mutation_prepared_sets_cache_mutex);
+
+    /// Cleanup stale entries where the shared_ptr is expired.
+    while (!mutation_prepared_sets_cache.empty())
+    {
+        auto it = mutation_prepared_sets_cache.begin();
```

Comment:
Do we rely on sequential mutations ids here? Are it always sequential? If yes, I can image onle case, we will not cleanup map if any mutation is much longer than other ones or it stuck. But most likely it's not really issue because size of this map relatively small.

Reply:
Yes, mutation ids are sequential. Theoretically there might be a case when the map is not cleaned up, but it is hard to imagine that is consumes much memory, because it only stores weak_ptr-s to cache objects, but the cache objects themselves are cleaned up when there are no mutation tasks referencing them.